src/client/auth/mod.rs

```diff
@@ -349,6 +407,42 @@ impl Credential {
     }
 }
 
+/// Contains the first client message sent as part of the authentication handshake.
+pub(crate) enum ClientFirst {
+    Scram(ScramVersion, scram::ClientFirst),
+    X509(Command),
+}
+
+impl ClientFirst {
+    pub(crate) fn to_document(&self) -> Document {
+        match self {
+            Self::Scram(version, client_first) => client_first.to_command(&version).body,
+            Self::X509(command) => command.body.clone(),
+        }
+    }
+
+    pub(crate) fn into_first_round(self, server_first: CommandResponse) -> FirstRound {
+        match self {
+            Self::Scram(version, client_first) => FirstRound::Scram(
+                version,
+                scram::FirstRound {
+                    client_first,
+                    server_first,
+                },
+            ),
+            Self::X509(..) => FirstRound::X509(server_first),
+        }
+    }
+}
+
+/// Contains the complete first round of the authentication handshake, including the client message
+/// and the server response.
+#[derive(Debug)]
+pub(crate) enum FirstRound {
+    Scram(ScramVersion, scram::FirstRound),
+    X509(CommandResponse),
```

Comment:
That logic can still remain internal to each of the authentication submodules though. I don't think invoking `ServerFirst::parse(doc)` is really any different than invoking `ClientFirst::to_command` (which we do in `set_speculative_auth_info`) in terms of caring about structure or breaking encapsulation.

Reply:
I think parsing the response is different because it's fallible; it seems a little strange to be failing in the handshake due to the shape of the response rather than in the regular auth flow.

Reply:
Ok, well at any rate, a  `Document` would also be fine by me here. I just wanted to avoid "emulating" a `CommandResponse`.

Reply:
Fair enough; I'll change it to Document